### PR TITLE
Iterate on the Notification Flow

### PR DESCRIPTION
**Pull request by dbuscaglia**
### Overview

Let's be honest. The original integration flow was kinda...(none) thangs. This new flow is legit. The system handles both global Alert notifications, swipe to refresh marks all as read (asynchronously), they are persisted and sorted from the backend, the receiver creates a broadcasted intent for any interested parties, all notification types are supported, and the system is generally robust. 
### Technical changes
- Sorted order by timestamp
- Allows for global app updates as well as in app generated notifications
- Clean builder pattern that meshes with parse
- Swipe to refresh grabs new notifications
### Reviewers
- [ ] @jcomo  
- [x] @hashamali

### CODE REVIEW

File: app/src/main/java/it/quip/android/fragment/NotificationsFragment.java [hunk 1]
```diff
@@ -40,9 +51,32 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
...
+        mNotifications.clear();
+        mNotifications.addAll(notifications);
+        mNotificationAdapter.notifyDataSetChanged();
+        mSwipeContainer.setRefreshing(false);
+    }
+
+
+    @Override
+    public void onException(ParseException e) {
+        Toast.makeText(this.getContext(), e.toString(), Toast.LENGTH_SHORT).show();
```
Comment (hashamali) [inline app/src/main/java/it/quip/android/fragment/NotificationsFragment.java]: Let's change this to a debug log of the error.

File: app/src/main/java/it/quip/android/fragment/NotificationsFragment.java [hunk 2]
```diff
@@ -40,9 +51,32 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
...
+        // None thangs
+    }
+
     private void attachNotificationAdapter(View view) {
         mRvContacts = (RecyclerView) view.findViewById(R.id.rv_notifications);
-        mNotificationAdapter = new NotificationAdapter(Notification.getNotifcations(0), this, this.getContext());
+        mNotifications = new ArrayList<>();
+        mNotificationAdapter = new NotificationAdapter(mNotifications
+                , this, this.getContext());
```
Comment (hashamali) [inline app/src/main/java/it/quip/android/fragment/NotificationsFragment.java]: no need for the line split here.

File: app/src/main/java/it/quip/android/model/Notification.java [hunk 3]
```diff
@@ -7,79 +9,213 @@
...
-    public void setViewed(boolean viewed) {
-        this.viewed = viewed;
+    public void setChannel(String channel) {
+        this.safePutKey(PUSH_CHANNEL_KEY, channel);
+        this.mChannel = channel;
     }
 
-    public long getUid() {
-        return uid;
+    public boolean viewed() {
```
Comment (hashamali) [inline app/src/main/java/it/quip/android/model/Notification.java]: `getViewed` to keep it consisted with other getters.

File: app/src/main/java/it/quip/android/model/Notification.java [hunk 4]
```diff
@@ -7,79 +9,213 @@
...
+    private boolean mViewed;
+
+    public String getChannel() {
+        mChannel = this.getString(PUSH_CHANNEL_KEY);
+        return mChannel;
     }
 
-    public void setViewed(boolean viewed) {
-        this.viewed = viewed;
+    public void setChannel(String channel) {
```
Comment (hashamali) [inline app/src/main/java/it/quip/android/model/Notification.java]: Can you move this setter down with the others.

File: app/src/main/java/it/quip/android/receiver/NotificationReceiver.java [hunk 5]
```diff
@@ -48,11 +52,25 @@ private void processPush(Context context, Intent intent) {
...
+                    // This is a global push, just use alert
+                    Notification notification = new Notification();
+                    notification.setTimestamp(TimeUtils.currentTimestampInS());
+                    notification.setReceiverUid(User.getUserForSession().getObjectId());
+                    notification.setText(json.getString("alert"));
+                    // TODO: use our image app for this?
+                    triggerBroadcastToActivity(context, notification);
+                }
+
                 Log.d(TAG, "got action " + action + " on channel " + channel + " with:");
```
Comment (dbuscaglia) [inline app/src/main/java/it/quip/android/receiver/NotificationReceiver.java]: Removing this crap

File: app/src/main/java/it/quip/android/model/Notification.java [hunk 6]
```diff
@@ -7,79 +9,213 @@
...
+        this.safePutKey(Notification.PUSH_IMAGE_URL_KEY, mNotificationImageUrl);
+    }
 
-        this.uid = uid;
-        this.text = text;
-        this.notificationImageUrl = notificationImageUrl;
-        this.timestamp = timestamp;
-        this.viewed = viewed;
-        this.type = STANDARD_NOTIFICATION;
+    private void safePutKey(String key, Object value) {
```
Comment (hashamali) [inline app/src/main/java/it/quip/android/model/Notification.java:206]: Let's move this into a shared BaseObject that all of our Parse model objects extend.

File: app/src/main/java/it/quip/android/model/Notification.java (hunk 3 above)
Comment (dbuscaglia) [inline app/src/main/java/it/quip/android/model/Notification.java]: viewed is non java its very pythonic, but reads so much better.  still will change

File: app/src/main/java/it/quip/android/model/Notification.java (hunk 4 above)
Comment (dbuscaglia) [inline app/src/main/java/it/quip/android/model/Notification.java]: yes

File: app/src/main/java/it/quip/android/adapter/NotificationBaseViewHolder.java [hunk 7]
```diff
@@ -43,6 +45,20 @@ public void onClick(View v) {
...
         });
 
+        timestampText.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+
+                Context context = mAdapter.getContext();
+                Notification notification = new Notification.with(context)
+                        .body("Jean Claude Van Como deposited some darkness in @sfsewers thangs")
+                        .sender(User.getUserForSession())
```
Comment (jcomo) [inline app/src/main/java/it/quip/android/adapter/NotificationBaseViewHolder.java:55]: This shoud be `QuipitApplication.getCurrentUser()`. `getUserForSession` is blocking and is only used right after the login

File: app/src/main/res/values/secrets.xml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="parse_client_key">CYgHSQKeDfLFu0i9eT7Ks2e8DPdgwTdOFEX2BLP8</string>
```
Comment (jcomo) [inline app/src/main/res/values/secrets.xml:3]: Yeah lets go ahead and remove this and not add it to the index

Comment (hashamali) [inline app/src/main/res/values/secrets.xml:3]: thangs

File: app/src/main/java/it/quip/android/adapter/NotificationBaseViewHolder.java (hunk 7 above)
Comment (hashamali) [inline app/src/main/java/it/quip/android/adapter/NotificationBaseViewHolder.java:55]: I've updated this in my branch.

File: app/src/main/res/values/secrets.xml (hunk 8 above)
Comment (dbuscaglia) [inline app/src/main/res/values/secrets.xml:3]: lol

**Pull request merged**